rewindable-index/src/RewindableIndex/Storable.hs

```diff
@@ -0,0 +1,358 @@
+module RewindableIndex.Storable
+  ( -- * State
+    Config
+  , memoryBufferSize
+  , diskBufferSize
+  , State
+  , handle
+  , config
+  , emptyState
+  , Storage
+  , storage
+  , events
+  , cursor
+  , getMemoryEvents
+  , getEvents
+  , filterWithQueryInterval
+  , StorableEvent
+  , StorablePoint
+  , StorableQuery
+  , StorableResult
+    -- * API
+  , QueryInterval(..)
+  , Buffered(..)
+  , Queryable(..)
+  , Resumable(..)
+  , Rewindable(..)
+  , HasPoint(..)
+  , insert
+  , insertMany
+  , rewind
+  , resume
+  , query
+  ) where
+
+import Control.Applicative ((<|>))
+import Control.Lens.Operators ((%~), (.~), (^.))
+import Control.Lens.TH qualified as Lens
+import Control.Monad.Primitive (PrimMonad, PrimState)
+import Data.Foldable (foldlM)
+import Data.Function ((&))
+import Data.Functor ((<&>))
+import Data.Vector qualified as V
+import Data.Vector.Generic qualified as VG
+import Data.Vector.Mutable qualified as VM
+import GHC.Generics (Generic)
+
+{-
+   The extensible parts of the indexers are the way data is stored into some form
+   of persistent storage. The interface used for working blockchain events is defined
+   in this module.
+
+   There are quite a bit of type variables that are involed in defining this interface.
+   Some of them stand for the type of database connection (h), the type of events (e),
+   the type of points along the blockchain (denoted by p), the type of queries (q) and
+   results.
+
+   The following data families implement the observation that most of the variables
+   can be derived from the way information is stored in the database, so there is
+   quite a bit of convenience in reducing the number of type variables to just two.
+   One that stands for the database connection (which should be a newtype for each
+   indexer) and one for the monad in which the indexer runs (usually IO).
+-}
+data family StorableEvent h
+
+data family StorablePoint h
+
```

Comment:
We've discussed it already on Slack but yes, having one`StorablePoint` per indexer sounds like a bad idea if we want to work with several of them while being sure they have indexed the same thing.